### ISSUE TITLE
Implementing a probabilistic neural network using Synaptic

### ISSUE THREAD

**Issue by codywatts** [cazala/synaptic#97]
First of all, let me say that Synaptic is a very cool project, so thank you to @cazala and all contributors.

I am trying to implement a probabilistic neural network by following [this paper by Donald Specht](http://www.wi.hs-wismar.de/~cleve/vorl/projects/dm/ss13/PNN/Quellen/Specht_ProbabilisticNeuralNetworks.pdf). However, I'm new to Synaptic (and neural-networks in general) and so it's not obvious to me whether Synaptic supports this kind of network.

The paper says:

> Each pattern unit [...] forms a dot product of the input pattern vector X with a weight vector W<sub>i</sub>, Z<sub>i</sub> = X • W<sub>i</sub>, and then performs a nonlinear operation on Z<sub>i</sub> before outputting its activation level to the summation unit.

Therefore, it seems to me that the input to the pattern unit is processed in a very particular way. However, [the documentation on Neurons](https://github.com/cazala/synaptic/wiki/Neurons#activate) states:

> When a neuron activates, it computes its state from all its input connections and squashes it using its activation function, and returns the output (activation).

Unfortunately, I do not understand how the neuron "computes its state from all its input connections" nor if this behavior can be overridden to perform the dot-product described in Specht's paper.

At this point, I am unsure how to proceed. Is it possible to implement a PNN using Synaptic? And if so, can anyone suggest how it might be done?

**Comment by cazala** [comment]
@codywatts yup, that's exactly what synaptic does. When a neuron activates, it takes all their inputs (this is basically a vector of floating values, we can call it Y), multiplies each of them by their corresponding connection weight (this is another vector of floats, we can call it W) and all this values are add together to get the **state** (S) of the neuron,

so, basically, to get the state of the neuron J, you have to perform a dot product:

![formula](https://camo.githubusercontent.com/875a8f9cef6889b90dcd82752d12569456f385f0/687474703a2f2f6c617465782e636f6465636f67732e636f6d2f6769662e6c617465783f735f6a25323025334425323025354373756d5f25374269253744253230775f253742696a2537442e795f25374269253744)

Then this values is passed thru a squashing function to get the activation of the neuron:

![squash](https://camo.githubusercontent.com/198be5a2e3b4b1e57758d46e4c324ac2c22671bd/687474703a2f2f6c617465782e636f6465636f67732e636f6d2f6769662e6c617465783f795f6a253230253344253230665f6a253238535f6a253239)

Now this can be an input for the next layer of neurons, or, if this is the last layer in the network, it will be part of the output vector.

If you have **gated connections** (like in LSTMs) there's actually one more thing going on, because when calculating the state of the neuron, the input vector (Y) is multiplied by the weight vector W and also the gain vector G, so is the dot product of `Y . W . G`, but when there are no gated connections (as in a Multilayer Perceptron) the gain is a vector of 1's, so it is the same as just a dot product between `Y . W`
